### PR TITLE
Improve rate limit exception

### PR DESCRIPTION
**Pull request by oleurud**
A part of https://github.com/CartoDB/cartoframes/issues/562

We want to add the related data (the HTTP headers: 'Carto-Rate-Limit-Limit', 'Carto-Rate-Limit-Remaining', 'Retry-After', 'Carto-Rate-Limit-Reset') to the exception fired when a request to Maps/SQL API is rate limited.

### CODE REVIEW

File: carto/exceptions.py [hunk 1]
```diff
@@ -10,10 +10,27 @@
...
 
 class CartoException(Exception):
     """
     Any Exception produced by carto-python should be wrapped around this class
     """
     pass
+
+
+class CartoRateLimitException(CartoException):
+    def __init__(self, response):
```
Comment (alrocar) [inline carto/exceptions.py]: docs please:

- what's this class for
- what attributes it includes
- maybe a link to the help center with more info could be handy
- etc.

Also let's take a look at the doc folder (and generated docs). There's a section about [Exceptions](https://carto-python.readthedocs.io/en/latest/general_concepts.html#exceptions) we may want to update.

File: carto/exceptions.py [hunk 2]
```diff
@@ -10,10 +10,27 @@
...
+        super().__init__(response.text)
+        self.limit = int(response.headers['Carto-Rate-Limit-Limit'])
+        self.remaining = int(response.headers['Carto-Rate-Limit-Remaining'])
+        self.retryAfter = int(response.headers['Retry-After'])
+        self.reset = int(response.headers['Carto-Rate-Limit-Reset'])
+
+    @staticmethod
+    def isResponseRateLimited(response):
```
Comment (alrocar) [inline carto/exceptions.py]: docs please (as a rule of thumb let's try to document any public method)

File: carto/exceptions.py [hunk 3]
```diff
@@ -10,10 +10,27 @@
...
     """
     pass
+
+
+class CartoRateLimitException(CartoException):
+    def __init__(self, response):
+        super().__init__(response.text)
+        self.limit = int(response.headers['Carto-Rate-Limit-Limit'])
+        self.remaining = int(response.headers['Carto-Rate-Limit-Remaining'])
+        self.retryAfter = int(response.headers['Retry-After'])
```
Comment (alrocar) [inline carto/exceptions.py]: beware the camelCase notation in variables

File: carto/exceptions.py (hunk 2 above)
Comment (alrocar) [inline carto/exceptions.py]: A naming suggestion 👼  I'd rename it as `isRateLimited`

File: carto/auth.py [hunk 4]
```diff
@@ -137,12 +136,15 @@ def send(self, relative_path, http_method, **requests_args):
...
             http_method, requests_args = self.prepare_send(http_method, **requests_args)
 
-            return super(APIKeyAuthClient, self).send(relative_path,
-                                                      http_method,
-                                                      **requests_args)
+            response = super(APIKeyAuthClient, self).send(relative_path, http_method, **requests_args)
         except Exception as e:
             raise CartoException(e)
 
+        if CartoRateLimitException.isResponseRateLimited(response):
```
Comment (alrocar) [inline carto/auth.py]: I guess we'll have to add this logic to the `NonVerifiedAPIKeyAuthClient` as well (it's used mainly in on-premises, for the cases when there's no SSL certificate installed in the server)

File: carto/exceptions.py [hunk 5]
```diff
@@ -20,15 +20,38 @@ class CartoException(Exception):
...
         super().__init__(response.text)
         self.limit = int(response.headers['Carto-Rate-Limit-Limit'])
         self.remaining = int(response.headers['Carto-Rate-Limit-Remaining'])
-        self.retryAfter = int(response.headers['Retry-After'])
+        self.retry_after = int(response.headers['Retry-After'])
         self.reset = int(response.headers['Carto-Rate-Limit-Reset'])
 
     @staticmethod
-    def isResponseRateLimited(response):
+    def isRateLimited(response):
```
Comment (alrocar) [inline carto/exceptions.py]: forgot to mention this method should not be camelCase 😛

Comment (alrocar) [reply carto/exceptions.py]: Looks like we could include [a plugin](https://github.com/PyCQA/pep8-naming) to detect this kind of things

Comment (oleurud) [reply carto/exceptions.py]: :ouch: I can't see them!

**Pull request merged**